File: common/src/main/java/cc/turtl/turtlshell/util/format/ComponentUtils.java
package cc.turtl.turtlshell.util.format;

import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.ClickEvent;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Iterator;
import java.util.function.Function;

/**
 * Provides basic chat component formatting utils.
 */
public final class ComponentUtils {
    /**
     * Placeholder component for missing or null data.
     */
    public static final Component UNKNOWN = createComponent("???", ColorUtils.DARK_GRAY.getRGB());
    public static final Component NONE = createComponent("None", ColorUtils.DARK_GRAY.getRGB());
    public static final Component SPACE = Component.literal(" ");
    public static final Component RESET = Component.literal("").withStyle(ChatFormatting.RESET);

    private ComponentUtils() {
    }

    public static MutableComponent createComponent(Object text) {
        return createComponent(text, ColorUtils.WHITE.getRGB());
    }

    /**
     * Creates a component with a specific color.
     * <p>Example: {@code literal("Lvl 50", ColorUtil.GOLD)}</p>
     */
    public static MutableComponent createComponent(Object text, int color) {
        return createComponent(text, color, false);
    }

    public static MutableComponent createComponent(Object text, int color, boolean bold) {
        String content = text == null ? "" : text.toString();

        MutableComponent component = Component.literal(content);

        return component.withStyle(style -> {
            style = style.withColor(color).withBold(bold);
            return style;
        });
    }

    /**
     * Creates a clickable URL component that opens the link in the browser on click.
     */
    public static Component clickableUrl(String url) {
        return Component.literal(url)
                .withStyle(s -> s
                        .withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_URL, url))
                        .withColor(ColorUtils.PINK.getRGB())
                        .withUnderlined(true));
    }

    /**
     * Creates a component from a label and value pair.
     * <p>Example: {@code label("Ability", "Intimidate")} -> "Ability: Intimidate"</p>
     */
    public static MutableComponent labelled(@NotNull Object label, @Nullable Object value) {
        MutableComponent labelComp = (label instanceof Component c ? c.copy() : Component.literal(label.toString()));
        // If we pass a component label, it gets overridden with gray for label consistency
        labelComp.withColor(ColorUtils.LIGHT_GRAY.getRGB());

        Component valueComp = (value == null) ? UNKNOWN :
                (value instanceof Component c ? c : createComponent(value.toString(), ColorUtils.WHITE.getRGB()));

        return labelComp.append(createComponent(": ", ColorUtils.LIGHT_GRAY.getRGB())).append(valueComp);
    }

    /**
     * Joins items into a single component with separators.
     * <p>Example: {@code join(list, ", ", item -> literal(item, ColorUtil.RED))}</p>
     */
    public static <E> Component join(@Nullable Iterable<E> items, String separator, Function<E, Component> mapper) {
        if (items == null || !items.iterator().hasNext()) return UNKNOWN;

        MutableComponent result = Component.empty();
        Iterator<E> it = items.iterator();

        while (it.hasNext()) {
            Component mapped = mapper.apply(it.next());
            if (mapped != null) {
                result.append(mapped);
                if (it.hasNext()) result.append(createComponent(separator, ColorUtils.DARK_GRAY.getRGB()));
            }
        }
        return result;
    }

    /**
     * Creates a component where the text is colored with a multi-point gradient.
     * * @param text   The string to color.
     *
     * @param colors The RGB color stops (0xRRGGBB).
     * @return A MutableComponent containing the gradient text.
     */
    public static MutableComponent gradient(String text, int... colors) {
        if (text == null || text.isEmpty()) return Component.empty();
        if (colors.length == 0) return Component.literal(text);
        if (colors.length == 1) return createComponent(text, colors[0]);

        MutableComponent result = Component.empty();
        int length = text.length();

        for (int i = 0; i < length; i++) {
            // Calculate ratio (0.0 to 1.0) based on character index
            // If length is 1, ratio is 0.
            float ratio = length > 1 ? (float) i / (length - 1) : 0.0f;

            int color = ColorUtils.getGradient(ratio, colors);
            result.append(Component.literal(String.valueOf(text.charAt(i))).withColor(color));
        }

        return result;
    }
}

File: common/src/main/kotlin/cc/turtl/turtlshell/util/format/ComponentUtils.kt
package cc.turtl.turtlshell.util.format

import net.minecraft.network.chat.ClickEvent
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.MutableComponent
import net.minecraft.network.chat.Style

val UNKNOWN: Component = Component.literal("???").withColor(ColorLib.DARK_GRAY.rgb)
val NONE: Component = Component.literal("None").withColor(ColorLib.DARK_GRAY.rgb)
val SPACE: Component = Component.literal(" ")

/**
 * Creates a clickable URL component.
 */
fun clickableUrl(url: String): MutableComponent =
    Component.literal(url).withStyle(
        Style.EMPTY
            .withClickEvent(ClickEvent(ClickEvent.Action.OPEN_URL, url))
            .withColor(ColorLib.PINK.rgb)
            .withUnderlined(true)
    )

fun labelled(
    label: Component,
    value: Component?,
    separatorColor: Int = ColorLib.LIGHT_GRAY.rgb,
    fallback: Component = UNKNOWN
): MutableComponent = label.copy()
    .append(Component.literal(": ").withColor(separatorColor))
    .append(value?.copy() ?: fallback.copy())

fun labelled(
    label: String,
    value: String?,
    labelColor: Int = ColorLib.LIGHT_GRAY.rgb,
    valueColor: Int = ColorLib.DARK_GRAY.rgb
): MutableComponent = labelled(
    Component.literal(label).withColor(labelColor),
    value?.let { Component.literal(it).withColor(valueColor) },
    labelColor
)

fun labelled(
    label: String,
    value: Component?,
    labelColor: Int = ColorLib.LIGHT_GRAY.rgb,
): MutableComponent = labelled(
    Component.literal(label).withColor(labelColor),
    value,
    labelColor
)

/**
 * Joins items into a single component with separators.
 *
 * Example: `joined(list, ", ") { literal(it, ColorUtil.RED) }`
 */
fun <E> joined(
    items: Iterable<E?>?,
    separator: String = ", ",
    separatorColor: Int = ColorLib.DARK_GRAY.rgb,
    mapper: (E?) -> Component?
): MutableComponent {
    val mapped = items?.mapNotNull { mapper(it) }?.takeIf { it.isNotEmpty() }
        ?: return UNKNOWN.copy()

    val sep = Component.literal(separator).withColor(separatorColor)
    return mapped.drop(1).fold(Component.empty().append(mapped.first())) { acc, c ->
        acc.append(sep.copy()).append(c)
    }
}

/**
 * Creates a component where the text is colored with a multipoint gradient.
 * @param text The string to color.
 * @param colors The RGB color stops (0xRRGGBB).
 * @return A MutableComponent with gradient-colored text.
 */
fun gradient(text: String?, vararg colors: Int): MutableComponent {
    if (text.isNullOrEmpty()) return Component.empty()
    if (colors.isEmpty()) return Component.literal(text)
    if (colors.size == 1) return Component.literal(text).withColor(colors[0])

    val result = Component.empty()

    text.forEachIndexed { i, char ->
        val ratio = i.toFloat() / (text.length - 1).coerceAtLeast(1)
        result.append(Component.literal(char.toString()).withColor(getGradient(ratio, *colors)))
    }

    return result
}